### PR TITLE
RSDK-3143 - implement cgo api calling into c for cartoFacade

### PR DESCRIPTION
**Pull request by kim-mishra**

### CODE REVIEW

File: viam-cartographer/src/carto_facade/carto_facade.h [hunk 1]
```diff
@@ -2,9 +2,13 @@
 #ifndef CARTO_FACADE_H
 #define CARTO_FACADE_H
 
+#ifdef __cplusplus
```
Comment (kim-mishra) [inline viam-cartographer/src/carto_facade/carto_facade.h:5]: @nicksanford I actually think these are C++ libraries, so they need to be within an #ifdef __cplusplus

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+type CViamCartoLib struct {
+	viam_carto_lib *C.viam_carto_lib
+}
+
+func NewViamCartoLib(miniloglevel int, verbose int) (*CViamCartoLib, error) {
+	pVcl := C.alloc_viam_carto_lib()
+	defer C.free_alloc_viam_carto_lib(pVcl)
+
+	ppVcl := (**C.viam_carto_lib)(unsafe.Pointer(pVcl))
```
[the remark this answers is not in the document]
Comment (kim-mishra) [reply viam-cartographer/src/carto_facade_go/carto_facade.go:42]: TODO: run `go vet` manually and confirm it passes 
TODO: check if `go vet` runs in CI and if not create a ticket to add it to CI

File: viam-cartographer.go [hunk 3]
```diff
@@ -28,11 +28,18 @@ import (
...
+	cfq "github.com/viamrobotics/viam-cartographer/viam-cartographer/src/carto_facade_queue"
 )
 
 // Model is the model name of cartographer.
 var Model = resource.NewModel("viam", "slam", "cartographer")
 
+// ViamCartoLib instantiates the c viam_carto_lib struct and provides global vars from the lib such as the logger.
+// TODO: instantiate viam_carto_lib with correct loglevels and verbosity
+// Question: what to do with this error?
+var ViamCartoLib, _ = cartoFacade.NewViamCartoLib(1, 1)
```
Comment (kim-mishra) [inline viam-cartographer.go:41]: var ViamCartoLib TYPE_OF_THIS_THING

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
+// Package cartofacade provides an api to call into c code
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:1]: Lets change the directory structure a bit:

1. lets have all the code your created under a package called `cartofacade` 
2. this file should be moved (with the same content) to `cartofacade/cgo_api.go`

File: viam-cartographer/src/carto_facade_queue/carto_facade_queue.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
+// Package cartofacadequeue defines all necessary components to enforce one call into c per cartofacade.
+package cartofacadequeue
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_queue/carto_facade_queue.go:2]: Lets change the directory structure a bit:

1. lets have all the code your created under a package called cartofacade
2. this file should be moved (with the same content) to cartofacade/queue.go

File: viam-cartographer.go [hunk 6]
```diff
@@ -177,6 +186,27 @@ func New(
 	return cartoSvc, nil
 }
 
+func setUpCartoFacadeQueue(cartoFacadeCancelCtx context.Context, activeBackgroundWorkers *sync.WaitGroup) cfq.CartoFacadeQueue {
```
Comment (nicksanford) [inline viam-cartographer.go]: Lets move this into viam-cartographer/src/carto_facade_queue/carto_facade_queue.go & make it public

File: viam-cartographer.go [hunk 7]
```diff
@@ -28,11 +28,16 @@ import (
 	dim2d "github.com/viamrobotics/viam-cartographer/internal/dim-2d"
 	"github.com/viamrobotics/viam-cartographer/sensors/lidar"
 	vcUtils "github.com/viamrobotics/viam-cartographer/utils"
+	cartoFacade "github.com/viamrobotics/viam-cartographer/viam-cartographer/src/carto_facade_go"
```
Comment (nicksanford) [inline viam-cartographer.go]: lets remove all changes to viam-cartographer.go

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+func successfulStatus(status C.int) bool {
+	// return int(status) == int(C.VIAM_CARTO_SUCCESS)
+	return true
+}
+
+// CViamCartoLib is a struct to hold the c type viam_carto_lib
+type CViamCartoLib struct {
+	viamCartolib *C.viam_carto_lib
+}
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:41]:
```suggestion
type CartoLib struct {
	value *C.viam_carto_lib
}
```

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+		// TODO: how to make this error useful?
+		return errors.New(string(status))
+	}
+	return nil
+}
+
+// CViamCarto is a struct to hold the c type viam_carto
+type CViamCarto struct {
+	viamCarto *C.viam_carto
+}
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:74]:
```suggestion
type Carto struct {
	value *C.viam_carto
}
```

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+// NewViamCartoLib calls viam_carto_lib_init and returns a pointer to a viam carto lib object.
+func NewViamCartoLib(miniloglevel, verbose int) (CViamCartoLib, error) {
+	pVcl := C.alloc_viam_carto_lib()
+	defer C.free_alloc_viam_carto_lib(pVcl)
+
+	ppVcl := (**C.viam_carto_lib)(unsafe.Pointer(pVcl))
+
+	status := C.viam_carto_lib_init(ppVcl, C.int(miniloglevel), C.int(verbose))
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:52]: Please write a function in this same file that converts all the error status codes (other than success) to go errors, who's message is the same as the name of the status.  Return the error derived from that function.

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+// CViamCartoLib is a struct to hold the c type viam_carto_lib
+type CViamCartoLib struct {
+	viamCartolib *C.viam_carto_lib
+}
+
+// NewViamCartoLib calls viam_carto_lib_init and returns a pointer to a viam carto lib object.
+func NewViamCartoLib(miniloglevel, verbose int) (CViamCartoLib, error) {
+	pVcl := C.alloc_viam_carto_lib()
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:45]: please add comments for why we need to do this

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+	defer C.free_alloc_viam_carto_lib(pVcl)
+
+	ppVcl := (**C.viam_carto_lib)(unsafe.Pointer(pVcl))
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:48]: 1. What does this do if `pVcl` is `nil` (which happens if malloc can't allocate memory)? Do we need to check for nil before passing it it to `unsafe.Pointer`?
2.  What happens if we don't do the cast?
3. Please comment why we need to call unsafe.Pointer

File: viam-cartographer/src/carto_facade_go/carto_facade.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicksanford) [reply viam-cartographer/src/carto_facade_go/carto_facade.go:48]: what was the answer to the todos?

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+
+	status := C.viam_carto_lib_init(ppVcl, C.int(miniloglevel), C.int(verbose))
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return CViamCartoLib{}, fmt.Errorf("error initializing viam_carto_lib status = %d", status)
+	}
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:54]:
```suggestion
	if err := statusErr(status), err != nil {
		return CViamCartoLib{}, err
	}
```

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+
+	vcl := CViamCartoLib{viamCartolib: (*ppVcl)}
+
+	return vcl, nil
+}
+
+// TerminateViamCartoLib calls viam_carto_lib_terminate to clean up memory for viam carto lib.
+func (vcl *CViamCartoLib) TerminateViamCartoLib() error {
+	status := C.viam_carto_lib_terminate((**C.viam_carto_lib)(unsafe.Pointer(vcl)))
+	if !successfulStatus(status) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:64]: please replace with `statusErr(status)`

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+	return nil
+}
+
+// CViamCarto is a struct to hold the c type viam_carto
+type CViamCarto struct {
+	viamCarto *C.viam_carto
+}
+
+// TODO: properly pass config
+func getConfig() C.viam_carto_config {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:77]: lets actually build out this function such that it is usable i.e. it takes a config, validates & if valid, returns the C type.

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+
+	return vcc
+}
+
+func cleanConfig(vcc C.viam_carto_config) {
+	C.free_bstring_array(vcc.sensors)
+}
+
+// NewViamCarto calls viam_carto_init and returns a pointer to a viam carto object.
+func NewViamCarto(vcl CViamCartoLib) (CViamCarto, error) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:102]: Lets have this function actually take the viam-cartographer config & convert it into the appropriate C type

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+	// TODO: get vcl as a param & initialize c queue channel here
+	pVc := C.alloc_viam_carto()
+	defer C.free_alloc_viam_carto(pVc)
+
+	ppVc := (**C.viam_carto)(unsafe.Pointer(pVc))
+
+	// TODO: Init with data
+	vcc := getConfig()
+	defer cleanConfig(vcc)
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:111]:
```suggestion
	defer C.free_bstring_array(vcc.sensors)
```

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+
+	componentReference := C.bfromcstr(C.CString("comp_ref"))
+	vcc.component_reference = componentReference
+
+	return vcc
+}
+
+func cleanConfig(vcc C.viam_carto_config) {
+	C.free_bstring_array(vcc.sensors)
+}
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:99]:
```suggestion
```

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+
+	ppVc := (**C.viam_carto)(unsafe.Pointer(pVc))
+
+	// TODO: Init with data
+	vcc := getConfig()
+	defer cleanConfig(vcc)
+	vcac := C.viam_carto_algo_config{}
+
+	status := C.viam_carto_init(ppVc, vcl.viamCartolib, vcc, vcac)
+	if !successfulStatus(status) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:115]: please replace with `statusErr(status)`

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+
+	vc := CViamCarto{viamCarto: (*ppVc)}
+	return vc, nil
+}
+
+// Start is a wrapper for viam_carto_start
+func (vc *CViamCarto) Start(ctx context.Context) error {
+	status := C.viam_carto_start(vc.viamCarto)
+
+	if !successfulStatus(status) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:128]: please replace with `statusErr(status)`

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+	}
+
+	return nil
+}
+
+// Stop is a wrapper for viam_carto_stop
+func (vc *CViamCarto) Stop(ctx context.Context) error {
+	status := C.viam_carto_stop(vc.viamCarto)
+
+	if !successfulStatus(status) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:140]: please replace with `statusErr(status)`

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+	}
+
+	return nil
+}
+
+// TerminateViamCarto calls viam_carto_terminate to clean up memory for viam carto
+func (vc *CViamCarto) TerminateViamCarto(ctx context.Context) error {
+	status := C.viam_carto_terminate(&vc.viamCarto)
+
+	if !successfulStatus(status) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:152]: please replace with `statusErr(status)`

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+	return nil
+}
+
+// CViamCartoAddSensorReading is a struct to hold the c type viam_carto_sensor_reading
+type CViamCartoAddSensorReading struct {
+	vcasr C.viam_carto_sensor_reading
+}
+
+// AddSensorReading is a wrapper for viam_carto_add_sensor_reading
+func (vc *CViamCarto) AddSensorReading(ctx context.Context) error {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:166]: Lets have this function actually take a lidar reading (probably a byte slice) & a timestamp & convert it into the appropriate C type.

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+	resp := CViamCartoAddSensorReading{vcasr: C.viam_carto_sensor_reading{}}
+
+	status := C.viam_carto_add_sensor_reading(vc.viamCarto, &resp.vcasr)
+
+	if !successfulStatus(status) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:171]: please replace with `statusErr(status)`

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 25]
```diff
@@ -0,0 +1,278 @@
...
+		// TODO: how to make this error useful?
+		return errors.New(string(status))
+	}
+
+	status = C.viam_carto_add_sensor_reading_destroy(&resp.vcasr)
+	if !successfulStatus(status) {
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:177]: please replace with `statusErr(status)`

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 26]
```diff
@@ -0,0 +1,278 @@
...
+		return errors.New(string(status))
+	}
+
+	return nil
+}
+
+// CViamCartoGetPositionResponse is a struct to hold the c type viam_carto_get_position_response
+type CViamCartoGetPositionResponse struct {
+	vcgpr C.viam_carto_get_position_response
+}
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:188]:
```suggestion
type CartoGetPositionResponse struct {
	value C.viam_carto_get_position_response
}
```

Comment (nicksanford) [reply viam-cartographer/src/carto_facade_go/carto_facade.go:188]: Also lets move all the struct definitions to the top of the file

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 27]
```diff
@@ -0,0 +1,278 @@
...
+		OZ: float64(resp.vcgpr.o_z),
+	}
+
+	status = C.viam_carto_get_position_response_destroy(&resp.vcgpr)
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return nil, "", errors.New(string(status))
+	}
+
+	// TODO: fill component reference in?
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:226]: Can we please do this todo?

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 28]
```diff
@@ -0,0 +1,278 @@
...
+	return spatialmath.NewPose(pos, ori), "", nil
+}
+
+// CViamCartoGetPointCloudMapResponse is a struct to hold the c type viam_carto_get_point_cloud_map_response
+type CViamCartoGetPointCloudMapResponse struct {
+	vcgpcmr C.viam_carto_get_point_cloud_map_response
+}
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:233]:
```suggestion
type CartoGetPointCloudMapResponse struct {
	value C.viam_carto_get_point_cloud_map_response
}
```

Lets also move this to the top of the file with the other structs

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 29]
```diff
@@ -0,0 +1,278 @@
...
+		return nil, errors.New(string(status))
+	}
+
+	return nil, errors.New("unimplemented error")
+}
+
+// CViamCartoGetInternalStateResponse is a struct to hold the c type viam_carto_get_internal_state_response
+type CViamCartoGetInternalStateResponse struct {
+	vcgisr C.viam_carto_get_internal_state_response
+}
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:258]:
```suggestion
type CartoGetInternalStateResponse struct {
	value C.viam_carto_get_internal_state_response
}
```
Lets move this to the top of the file with the rest of the structs

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 30]
```diff
@@ -0,0 +1,278 @@
...
+		return nil, errors.New(string(status))
+	}
+
+	status = C.viam_carto_get_internal_state_response_destroy(&resp.vcgisr)
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return nil, errors.New(string(status))
+	}
+
+	return nil, errors.New("unimplemented error")
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:277]: Can we actually implement this?

File: viam-cartographer/src/carto_facade_go/carto_facade.go [hunk 31]
```diff
@@ -0,0 +1,278 @@
...
+		return nil, errors.New(string(status))
+	}
+
+	status = C.viam_carto_get_point_cloud_map_response_destroy(&resp.vcgpcmr)
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return nil, errors.New(string(status))
+	}
+
+	return nil, errors.New("unimplemented error")
```
Comment (nicksanford) [inline viam-cartographer/src/carto_facade_go/carto_facade.go:252]: Can we actually implement this?

**Review by nicksanford — CHANGES_REQUESTED** [review]

**Pull request closed without merging**